### DISCUSSION TITLE
print table contents

### DISCUSSION THREAD

**Discussion by lazyclaw-labs** [diesel-rs/diesel#3248] [Ideas]
Is there a way to use Diesel to print the content of a table to screen, without the data type knowledge?

i.e.: Print("select top 1 * from anysqlservertable");
The format could be CSV, or whatever. 

This is very basic in other platform, but somehow it seems almost impossible in rust.

**Comment by lazyclaw-labs** [comment]
or any mysql or postgres table. Simply dump the content to screen. Is it possible?

**Reply by weiznich** [reply]
It's possible to implement something like this on top of diesel (or better `diesel_dynamic_schema`), but it requires a number of assumptions about types that can occur and a specifically tailored implementation.
Diesel does not support such things out of the box because such a query can return a unrestricted set of types, where we would need to know for each type how to parse it from database representation and how print it as CSV value. It's really not a "simple" thing and I doubt that anything beside sqlite's command line tool can support this in a generic complete way at all.

**Reply by lazyclaw-labs** [reply]
I understand it's not simple, or else, I won't be asking here. But the functionality is basic. Like you mention even a commandline tool has it. We wrote so much rusty code, can't even match a commandline tool?

I come from a .net world. worked with ADO.NET/C#. This kind of functionality comes right out of the box. I mean, you create a connection, execute a command, got a datareader, you read a value (Don't know it's datatype yet, you can certainly test its type though),, call ToString(option formatter) on the value, you got the string presentation of the value.
Suppose I want to build a tool like DBeaver, it's easy in .net world, but it seems impossible in RUST. 
I found that all the sql implementation, Diesel, SqlX, won't provide such a basic functionality.

In my opinion, all you need to do is to map all the possible SQLTypes of a database to a default rust type, and ensure every mapped rust type implement Display trait (most of the rust built-in type already have the Display trait implemented).
And this is supposed to be the base of all DB provider. Everything else, like ORM stuff, should be built above this.

It may seem a a lot of tedious work, but I thought that's what a framework is for, Get all the dirty work done for developers..

**Reply by weiznich** [reply]
There is a lot of misconception in your post:

* There is no such thing as "all the possible SQLTypes". This is a unbound set of types, which cannot be known beforehand, as users can provide custom type definitions as well
* There is no 1:1 mapping between database side types and rust types. As soon as you need to handle anything more complex as a simple number, there might be more than one matching type on rust side. Even things like `Text` fields can be mapped to more than one rust type. And that does not even mention types that just not have a counterpart in rusts standard library and where more than one crate exists to represent such a type
* I'm quite sure that all those "simple" command line tools contain quite a bit of custom code to implement this functionality. Diesel is no command line tool and optimizes for different use cases.
* Exposing such a functionality is by no means required for implementing a good performant database library. In contrary, it introduces another intermediator value representation, that the library needs to (de)serialize to.
* It's not true that it is impossible to build something like this using diesel. As already stated before it's just not something diesel is optimized for and that is documented in an accessible way. [`LoadConnection::load`](https://docs.diesel.rs/master/diesel/connection/trait.LoadConnection.html#tymethod.load) exposes an iterator over database rows. These rows then can be used to access individual fields. Each field includes information about the field name and run-time type information. You "only" need to convert the database representation of the attached value to whatever you like.

With all that above written: Noone forces you to use diesel or rust. If you believe that another language or library is a better solution to your problem, why don't you just use that?

> It may seem a a lot of tedious work, but I thought that's what a framework is for, Get all the dirty work done for developers..

This sounds like that you expect that "all the hard" work is done by someone else for free. If that's the case, I want to be very clear: This behaviour is not welcome here. We don't implement large scale features just because some stranger shows up in our issue tracker and requested them to be implemented for free. Such statements are the best way to be ignored at all.

If you are genuinely interested in improving the current state here, I'm open to discuss concrete improvements on PR basis as long as they do not degrade performance or other use-cases diesel is used for.
